src/main/java/com/cloudbees/plugins/credentials/CredentialsProvider.java

```diff
@@ -971,7 +971,7 @@ public CredentialsStore next() {
      * @return the trigger of the supplied run or {@code null} if this could not be determined.
      */
     @CheckForNull
```

Comment:
```suggestion
    @Restricted(NoExternalUse.class)
    @CheckForNull
```

We do not expect this to be called from plugins, right? And the use of `Map.Entry` for a tuple is too ugly to bake into any supported API.

Reply:
Well, I think the problem here might be that the `build` step will also need to bind a credential parameter binding, so it needs to figure out who triggered that build before propagating it down. Or say we have a pipeline where I bind my user scoped credentials during an `input` step (or as a build parameter) which is needed later on in a `build` step which also takes a build parameter for the credential. The invoked job won't propagate this credential binding info stored in the outer job without implementing the feature I was thinking of for the `build` step. See https://github.com/jenkinsci/credentials-plugin/blob/master/docs/consumer.adoc#binding-user-supplied-credentials-parameters-to-builds for the official API I put together back when I added support to `input` and build parameters for shadowing credentials with user-scoped credentials.

Reply:
^^^ probably meant as a top-level topic, not a comment on this trivial suggestion?